README.md
# rammerhead

> proxy based on testcafe-hammerhead (password is `sharkie4life`)

Demo link: https://demo-opensource.rammerhead.org

Polished closed-source-for-now browser version: https://browser.rammerhead.org (more links by awesome community members in the discord server at the end of readme)


## Supporting me and contributing

Server infrastructure costs money and developing this project consumes a lot of my time, so I would appreciate it greatly if you become a Patreon member: https://www.patreon.com/rammerhead


## Who is this package for

Package is for those who want a fully-configurable proxy that works on many sites

## Effectiveness of proxy

This proxy supports proxying
- basically everything

## To sign in to a Google account...

You need to go through specific steps, else it will tell you that the browser is not secure.
It may ask for a CAPTCHA, you will only be able to listen to audio one.

1. In Rammerhead browser, open **classroom.google.com**
2. Then on **Sign in to Classroom**
3. A popup window will open, allowing you to sign in without any error
4. Google Classroom will now open, and your google account will be logged in. You can close Classroom, and I advise you to save the rammerhead backup to avoid the hassle of completing captcha.

## Features of proxy

The proxy allows users to create a "session". When they access their session, their localStorage and cookies will be synced with rammerhead. This allows for accurately mocking cookied requests and conveniently save their logins even if they switch devices. This also enables users to configure a custom HTTP proxy server for rammerhead to connect to for the session.

## Installing and running

Rammerhead recommends you to have at least **node v16** to be installed. Once you installed nodejs, clone the repo, then run `npm install` and `npm run build`.

After, configure your settings in [src/config.js](src/config.js). If you wish to consistently pull updates from this repo without the hassle of merging, create `config.js` in the root folder so they override the configs in `src/`.

Finally run the following to start rammerhead: `node src/server.js`

## Discord server

For any user-help non-issue related questions, especially pertaining to Rammerhead Browser, please ask them here: [Rammerhead Support Server](https://discord.gg/VNT4E7gN5Y).


CHANGELOG.md
## v1.2.63

- add ability to override proxy settings on server-level basis
- update `testcafe-hammerhead` from `24.5.18` to `31.6.2`

## v1.2.62

- fix throw error on empty files

## v1.2.61

- fix missing cors origin header when port and crossDomainPort are the same

## v1.2.6

- properly fixed window.top issues when window.top isn't hammerhead
- replace unreliable `win.__get$(win, 'location')` with just rewriting the url directly 
- fix worker-hammerhead.js proxy url rewriting port undefined when 443 or 80

## v1.2.51

- add newly added classes to exports in src/index.js

## v1.2.5

- replace broken `keyv-lru-files` with own implementation of JS caching

## v1.2.42

- fix localStorage's getItem referencing parent in iframes

## v1.2.41

- handle removeStaleSessions of .get() returning undefined from corrupted session files

## v1.2.4

- fix crashes from corrupted sessions

## v1.2.3

- fix memory usage issues when downloading huge files
- fix iframing cross-origin proxy

## v1.2.2

- add disk cache option for processed JS files. fixes huge server memory usage and enables workers to share the same cache
- update `testcafe-hammerhead` to `v24.5.18`. fixes huge server slowdowns as brotli compression level is now adjusted to a much more reasonable value

## v1.2.11

- fix huge spikes of memory usage by replacing localStorage system with a custom one
- more fixes for iframing

## v1.2.01

- avoid using unstable API `fs.cpSync` in build.js

## v1.2.0

- added multithreading support

## v1.1.34

- convert hooks to stackable rewrite system

## v1.1.33

- delete hooks only after all fix function calls

## v1.1.32

- fix localStorage communication between windows by forcing them to read/write from realLocalStorage on every (get/set)Item call 

## v1.1.31

- add argument for ignoring files in `addStaticFilesToProxy`
- fix parseProxyUrl().proxy.port for 443 and 80 urls

## v1.1.3

- add option to restrict IP to session

## v1.1.21

- fix rewriting only non-websocket server headers
- fix errors when calling focus()/click()... to a closed iframe
- don't strip headers (hook onto res.writeHead) if connection is a websocket

## v1.1.2

- build to rammerhead.js and rammerhead.min.js
- fix same-domain iframes
- add jsdoc definitions for rammerhead store classes
- fix http proxy setting not deleting correctly

## v1.1.1

- fix uncatchable connection crash errors
- avoid shuffling percent encodings
- prevent forwarding localStorage endpoint to site by referrer
- fix (un)shuffle for location.hash and location.search

## v1.1.0

- add url encoding
- handle ECONNRESET manually
- bring back MemoryStore class for module exports
- add server option to disable localStorage syncing
- fix `RammerheadSessionFileCache` not saving cache to disk correctly

## v1.0.8

- handle websocket EPIPE error
- replace hammerhead's connection reset guard with a non-crashing rammerhead's reset guard
- add missing element attr getter unrewrite
- fix url rewriting for ports 80 and 443

## v1.0.7

- disable http2 support (for proxy to destination sites) because error handling is too complicated to handle
- removed server headers `report-to` (to avoid proxy url leak) and `cross-origin-embedder-policy` (which fixes reCAPTCHA v3)

## v1.0.61

- fix logger.error undefined (caused by not fully updating arguments for httpResponse.badRequest)

## v1.0.6

- expose more utils for npm package
- show password box if needed for html demo

## v1.0.5

- expose more modules for npm package
- add support for .env files
- add `deleteUnused` config option
- fix default 3 day session delete

## v1.0.43

- revert "revert fix for fix npm package"

## v1.0.42

- add entrypoint index.js for rammerhead package
- add package-lock.json to source control

## v1.0.41

- update demo link
- fix npm package

## v1.0.4

- add support for environment variable `DEVELOPMENT`
- fix crash when fetching /deletesession with a non-existent session id

## v1.0.3

- fix stability issues with websocket

## v1.0.2

- update `testcafe-hammerhead` to `v24.5.13`

## v1.0.1

- removed multi worker and rate limiting support to defer the complexity to other more suitable platforms like Docker. See [this commit](https://github.com/binary-person/rammerhead/tree/31ac3d23f30487f0dcd14323dc029f4ceb3b235a) if you wish to see the original attempt at this.
- removed unused session cleanup (as traversing the session list forces the cache into memory)
- lots of cleanup

## v1.0.0

- Initial commit


src/build.js
require('dotenv-flow').config();

const path = require('path');
const fs = require('fs');
const UglifyJS = require('uglify-js');

// modify unmodifable items that cannot be hooked in rammerhead.js
fs.writeFileSync(
    path.join(__dirname, './client/hammerhead.js'),
    fs
        .readFileSync(path.join(__dirname, '../node_modules/testcafe-hammerhead/lib/client/hammerhead.js'), 'utf8')
        // part of fix for iframing issue
        .replace('(function initHammerheadClient () {', '(function initHammerheadClient () {' +
            'if (window["%is-hammerhead%"]) throw new TypeError("already ran"); window["%is-hammerhead%"] = true;' +
            'window.rammerheadTop = (function() {var w = window; while (w !== w.top && w.parent["%hammerhead"]) w = w.parent; return w;})();' +
            'window.rammerheadParent = window.rammerheadTop === window ? window : window.parent;' +
            'window.distanceRammerheadTopToTop = (function() { var i=0,w=window; while (w !== window.top) {i++;w=w.parent} return i; })();' +
            'window.rammerheadAncestorOrigins = Array.from(location.ancestorOrigins).slice(0, -window.distanceRammerheadTopToTop);\n')
        // fix iframing proxy issue.
        // we replace window.top comparisons with the most upper window that's still a proxied page
        .replace(
            /(window|win|wnd|instance|opener|activeWindow)\.top/g,
            '$1.rammerheadTop'
        )
        .replace(
            /window\.parent/g,
            'window.rammerheadParent'
        )
        .replace(
            /window\.location\.ancestorOrigins/g,
            'window.rammerheadAncestorOrigins'
        )
        .replace(
            'isCrossDomainParent = parentLocationWrapper === parentWindow.location',
            'isCrossDomainParent = parentLocationWrapper === parentWindow.location || !parentWindow["%hammerhead%"]'
        )
        .replace(
            '!sameOriginCheck(window1Location, window2Location)',
            '!(sameOriginCheck(window1Location, window2Location) && (!!window1["%is-hammerhead%"] === !!window2["%is-hammerhead%"]))'
        )
        // return false when unable to convert properties on other windows to booleans (!)
        .replace(
            /!(parent|parentWindow|window1|window2|window\.top)\[("%(?:is-)?hammerhead%")]/g,
            '!(() => { try{ return $1[$2]; }catch(error){ return true } })()'
        )

        // disable saving to localStorage as we are using a completely different implementation
        .replace('saveToNativeStorage = function () {', 'saveToNativeStorage = function () {return;')

        // prevent calls to elements on a closed iframe
        .replace('dispatchEvent: function () {', '$& if (!window) return null;')
        .replace('click: function () {', '$& if (!window) return null;')
        .replace('setSelectionRange: function () {', '$& if (!window) return null;')
        .replace('select: function () {', '$& if (!window) return null;')
        .replace('focus: function () {', '$& if (!window) return null;')
        .replace('blur: function () {', '$& if (!window) return null;')
        .replace('preventDefault: function () {', '$& if (!window) return null;')

        // expose hooks for rammerhead.js
        .replace(
            'function parseProxyUrl$1',
            'window.overrideParseProxyUrl = function(rewrite) {parseProxyUrl$$1 = rewrite(parseProxyUrl$$1)}; $&'
        )
        .replace(
            'function getProxyUrl$1',
            'window.overrideGetProxyUrl = function(rewrite) {getProxyUrl$$1 = rewrite(getProxyUrl$$1)}; $&'
        )
        .replace('return window.location.search;', 'return (new URL(get$$2())).search;')
        .replace('return window.location.hash;', 'return (new URL(get$$2())).hash;')
        .replace(
            'setter: function (search) {',
            '$& var url = new URL(get$$2()); url.search = search; window.location = convertToProxyUrl(url.href); return search;'
        )
        .replace(
            'setter: function (hash) {',
            '$& var url = new URL(get$$2()); url.hash = hash; window.location.hash = (new URL(convertToProxyUrl(url.href))).hash; return hash;'
        )
        // sometimes, postMessage doesn't work as expected when
        // postMessage gets run/received in same window without hammerhead wrappings.
        // this is to double check hammerhead wrapped it
        // (cloudflare's turnsile threw this error after it tried to postMessage a fail code)
        .replace(
            'data.type !== MessageType.Service && isWindow(target)',
            '$& && data.type?.startsWith("hammerhead|")'
        )
);

// fix the
// worker-hammerhead.js:2434 Uncaught TypeError: Cannot read properties of undefined (reading 'toString')
//     at worker-hammerhead.js:2434:35
fs.writeFileSync(
    path.join(__dirname, './client/worker-hammerhead.js'),
    fs
        .readFileSync(path.join(__dirname, '../node_modules/testcafe-hammerhead/lib/client/worker-hammerhead.js'), 'utf8')
        .replace('proxyLocation.port.toString()', 'proxyLocation.port?.toString() || (proxyLocation.protocol === "https:" ? 443 : 80)')
);

// fix the
// transport-worker.js:1022 Uncaught TypeError: Cannot read properties of undefined (reading 'toString')
//     at transport-worker.js:1022:38
fs.writeFileSync(
    path.join(__dirname, './client/transport-worker.js'),
    fs
    .readFileSync(path.join(__dirname, '../node_modules/testcafe-hammerhead/lib/client/transport-worker.js'), 'utf8')
    .replace('proxyLocation.port.toString()', 'proxyLocation.port?.toString() || (proxyLocation.protocol === "https:" ? 443 : 80)')
);

const minify = (fileName, newFileName) => {
    const minified = UglifyJS.minify(fs.readFileSync(path.join(__dirname, './client', fileName), 'utf8'));
    if (minified.error) {
        throw minified.error;
    }
    fs.writeFileSync(path.join(__dirname, './client', newFileName), minified.code, 'utf8');
};

minify('rammerhead.js', 'rammerhead.min.js');
minify('hammerhead.js', 'hammerhead.min.js');
minify('worker-hammerhead.js', 'worker-hammerhead.min.js');
minify('transport-worker.js', 'transport-worker.min.js');
